src/main/java/com/coveo/spillway/storage/utils/SlidingCacheSynchronisation.java

```diff
@@ -0,0 +1,108 @@
+ * Copyright (c) 2016 Coveo
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.coveo.spillway.storage.utils;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.TimerTask;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.coveo.spillway.limit.LimitKey;
+import com.coveo.spillway.storage.LimitUsageStorage;
+import com.coveo.spillway.storage.sliding.AsyncSlidingLimitUsageStorage;
+import com.coveo.spillway.storage.sliding.InMemorySlidingStorage;
+
+/**
+ * Synchronization {@link TimerTask} that is launched periodically
+ * by the {@link AsyncSlidingLimitUsageStorage}.
+ *
+ * @author Emile Fugulin
+ * @since 1.1.0
+ */
+public class SlidingCacheSynchronisation extends TimerTask {
+  private static final Logger logger = LoggerFactory.getLogger(CacheSynchronization.class);
+
+  private InMemorySlidingStorage cache;
+  private LimitUsageStorage storage;
+  private Duration durationToSync;
+
+  public SlidingCacheSynchronisation(
+      InMemorySlidingStorage cache, LimitUsageStorage storage, Duration durationToSync) {
+    this.cache = cache;
+    this.storage = storage;
+    this.durationToSync = durationToSync;
+  }
+
+  @Override
+  public void run() {
+    Instant oldestKey = Instant.now().minus(durationToSync);
+
+    System.out.println("Now    : " + Instant.now().toEpochMilli());
+    System.out.println("Oldest : " + oldestKey.toEpochMilli());
+
+    cache.applyOnEach(
+        instantEntry -> {
+          try {
+            Instant bucket = instantEntry.getKey();
+
+            System.out.println("Current: " + bucket.toEpochMilli());
+
+            if (bucket.isBefore(oldestKey)) {
+              return;
+            }
+
+            System.out.println("Sync");
+
+            instantEntry
+                .getValue()
+                .entrySet()
+                .forEach(
+                    valueEntry -> {
```

Comment:
a lambda in a lambda, i'd split taht into different methods, else you end you in javascript callback-hell